### PR TITLE
New jnl

### LINKED ISSUE #1652: JNL lacking flexibility with different cargo
**Is your feature request related to a problem? Please describe.**
JNL currently is made for one cargo size and cant handle mounts and cargo being loaded onto a vehicle at the same time. it also has been left unmaintained.

**Describe the solution you'd like**
Remake to a more flexible system, fixing the bugs in the process.

### PR DESCRIPTION
**Pull request by HakonRydland**
## What type of PR is this.
1. [ ] Bug
2. [ ] Change
3. [x] Enhancement

### What have you changed and why?
Information: JNL replacement Logistics (aka newJNL)
JNL was full of bugs from lack of maintenance and extremely un-flexible, and with the common agreement of moving away from all the JN systems, there was no need to keep JNL in favor of another system as it no longer would have integration in a partially functional system anymore.

So with this in mind and a desire for a more flexible logistics system; voiced by the community during testing of LTC, I decided to resolve this, as JNL was riddled with bugs and I needed to completely replace the core functionality; not necessarily heavily modify the surrounding logic. I decided to fully replace JNL with a new system.

The new system will be consisting of vehicles with x number of EQUAL nodes to handle both cargo and weapon loading; these node will be equal in size, the cargo would then “use” x amount of these nodes depending on the defined size of the cargo, from the smallest “surrender crates” with a size of 1 to the large; and realistically unloadable in with the current defined vehicles if matching vehicles to mod set “ZU-23” with a size of 7 nodes.
Per request the new system features a blacklist for the weapons defined to block it from being mounted/loaded onto certain vehicles(models) for example mortars on assault boats, loading weapons in a covered transport vehicle or in the back of a closed van; this is done with a general weapon loading ban on covered/closed vehicles (manually defined).

To summarize JNL is unmaintained, buggy, and inflexible; to the point where a new system was the best option.

### Please specify which Issue this PR Resolves.
closes #1652 

### Please verify the following and ensure all checks are completed.

1. [x] Have you loaded the mission in singleplayer?
2. [x] Have you loaded the mission in LAN host?
3. [x] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [x] No
2. [ ] Yes (Please provide further detail below.)

### How can the changes be tested?
Steps: preferable dedicated server, use all available options of the system and generally try to break it, its only use will be trough the action added with ``A3A_fnc_logistics_addLoadAction;`` 

********************************************************
Notes: when we get around to doing MIE the only change will be how the nodeInit is done, as then it will be built in templates, then distributed by nodeInit from the server only, in contrast with now where they are run on each client locally.

### CODE REVIEW

**Review by wurzel0701 — COMMENTED** [review]
There are two different inputs for A3A_fnc_logistics_addLoadAction
surrenderAction.sqf
AIVEHinit.sqf
createAIAirplane.sqf
createAIOutpost.sqf
createAISite.sqf
use Object call A3A_....
while
flagAction.sqf
AIVEHinit.sqf
spawnCrate.sqf
use [object] call A3A_...
From the looks of it, the array is the right input, but maybe I am missing something

File: A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    Environment: Any
+    Public: [Yes]
+    Dependencies:
+
+    Example: [_object] call A3A_fnc_logistics_addLoadAction;
+*/
+params ["_object", ["_action", "load"]];
+
+if (isNil "logistics_vehicleHardpoints") exitWith {
+	[1, "Logistics nodes not initialized", "fn_logistics_addLoadAction.sqf"] call A3A_fnc_log;
```
Comment (wurzel0701) [inline A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf:23]: Wouldnt it be better to initialize the vehicle hardpoints instead of aborting?
Or is this not possible?

File: A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+};
+
+_nonSupportedStatic = false;
+if (_object isKindOf "StaticWeapon") then {
+	private _model = getText (configFile >> "CfgVehicles" >> typeOf _object >> "model");
+	if (({(_X#0) isEqualTo _model } count logistics_weapons) isEqualTo 0) then {_nonSupportedStatic = true};
```
Comment (wurzel0701) [inline A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf:29]: To not count the whole array, you can do
```logistics_weapons findIf ((_x#0) isEqualTo _model) == -1```
This will return -1 and result in true if nothing is found, however it will abort the search as soon as it finds the first matching element

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_addAction.sqf [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+*/
+params ["_object", "_action"];
+
+private _actions = [];
+{_actions pushBack (_object actionParams _x)}forEach actionIDs _object;
+private _loadText = format ["Load %1 into nearest vehicle", getText (configFile >> "CfgVehicles" >> typeOf _object >> "displayName")];
+
+switch (_action) do {
+    case "load":{
+        if (_actions findIf {(_x#0) isEqualTo _loadText} != -1) exitWith {};
```
Comment (wurzel0701) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_addAction.sqf:28]: I am not sure how fast sqf is with comparing texts (maybe the just hash it, but you know, BI...

I would add a variable to the object itself, either holding the load action or the load state

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_addOrRemoveObjectMass.sqf [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+		getText(configFile >> "cfgVehicles" >> typeOf _veh >> "displayName"),
+		_vehDefaultMass,
+		_cargoMass,
+		_vehDefaultMass + _cargoMass
+	];
+	_text = _text + _msg;
+	
+	//Don't use customHint, as we've got a lot of custom formatting.
+	[parseText _text] remoteExec ["hint", remoteExecutedOwner];
+	[parseText _text] remoteExec ["hint", crew _vehicle]; //parse text dont work with chat
```
Comment (wurzel0701) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_addOrRemoveObjectMass.sqf:44]: From my experience most vehicles are empty when you load/unload stuff, personally I have never seen that message. A search for nearby vehicles might be a good idea

Comment (HakonRydland) [reply A3-Antistasi/functions/Logistics/functions/fn_logistics_addOrRemoveObjectMass.sqf:44]: can you expand on the search for nearby vehicles part, as for the hint crew, i definitely have before been in example driver while someone else loads the crate, regardless this is actually one that is virtually untouched from the original, only containing a fix for the chat message being done with parse text, this is the now hint to the crew.

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_addAction.sqf (hunk 3 above)
Comment (HakonRydland) [reply A3-Antistasi/functions/Logistics/functions/fn_logistics_addAction.sqf:28]: a set/getVariable would be a bit faster, but not drastically, and with how few times this will actually be run relatively i dont find the extra clutter in the players/vehicles variable to be worth it.

File: A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf (hunk 2 above)
Comment (HakonRydland) [reply A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf:29]: done

File: A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf (hunk 1 above)
Comment (HakonRydland) [reply A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf:23]: the nodes are defined during initClient currently, when we move to MIE it will be defined in templates/initServer then publicVariabled, in either case they should be defined long before we try to add a action to any object, and for JIP it dosnt need the data yet. i have not encountered a single instance of it having to abort as of yet. however if youd like i recommend looking for the logged line in Cinders server logs to verify, there should be more than enough data there to be certain.

TL;DR it dosnt even actually need that line, it was left in to find out if it ever would actually would fail, which it hasnt afaik.

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_addOrRemoveObjectMass.sqf (hunk 4 above)
Comment (CalebSerafin) [reply A3-Antistasi/functions/Logistics/functions/fn_logistics_addOrRemoveObjectMass.sqf:44]: customHint takes precompiled text. If you do this, the client will never see the hint there.

Comment (CalebSerafin) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_addOrRemoveObjectMass.sqf:44]:
```suggestion
	["logistics_objtMass_owner",parseText _text] remoteExec ["A3A_fnc_customHint", remoteExecutedOwner];
	["logistics_objtMass_crew",parseText _text] remoteExec ["A3A_fnc_customHint", crew _vehicle];
```
you can change the header text, as in this case it is only used as an identifier.

**Review by CalebSerafin — CHANGES_REQUESTED** [review]
Could `A3-Antistasi/functions/Logistics/adding stuff to the new JNL.docx` have its content moved to a simple text file? It would drastically reduce the file size, and be viewable in all code editors.

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_initMountedWeapon.sqf [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+	private _location = _weapon getVariable ["AttachmentOffset", [0,0,0]];
+
+	private _appliedForce = (_fireDir vectorMultiply -_force);
+	_vehicle addForce [_appliedForce, _location];
+}];
+
+[_weapon, _idRecoil] spawn {
+	params ["_weapon", "_idRecoil"];
+	waitUntil {sleep 1; (attachedTo _weapon) isEqualTo objNull};
+	_weapon removeEventHandler ["Fired", _idRecoil];
```
Comment (CalebSerafin) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_initMountedWeapon.sqf:47]: These lines are indented with tabs.

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_generateNodes.sqf [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		_adjustedArray pushBack _element;
+	} forEach _nodeArray;
+
+	_nodeArray = _adjustedArray;
+};
+
+//fix nodeArray with rest of information
+private _newArray = [];
+{
+	_newArray pushBack [1, _x, []];
```
Comment (CalebSerafin) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_generateNodes.sqf:54]: This new file has Ln31 and Ln32 indented with spaces and the rest with tabs.

File: A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+};
+if (_nonSupportedStatic) exitWith {};
+
```
Comment (CalebSerafin) [inline A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf:32]: These lines are indented with tabs.

**Review by CalebSerafin — COMMENTED** [review]
Okay, what I flag was sorted. Converting Tabs to spaces in functions.hpp, I guess it was going to have to happen eventually, current PRs will just need to fix their minor merge conflicts if they removed/renamed a function.

**Review by CalebSerafin — APPROVED** [review]

File: A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+params ["_object", ["_action", "load"]];
+
+if (isNil "logistics_vehicleHardpoints") exitWith {
+    [1, "Logistics nodes not initialized", "fn_logistics_addLoadAction.sqf"] call A3A_fnc_log;
+};
+
+_nonSupportedStatic = false;
+if (_object isKindOf "StaticWeapon") then {
+    private _model = getText (configFile >> "CfgVehicles" >> typeOf _object >> "model");
+    if (logistics_weapons findIf ((_x#0) isEqualTo _model) isEqualTo -1) then {_nonSupportedStatic = true};
```
Comment (jaj22) [inline A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf:29]: Wrong bracket type on the findIf. Does this actually work?

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_getVehicleNodes.sqf [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+    Scope: Server,Server/HC,Clients,Any
+    Environment: Scheduled/unscheduled/Any
+    Public: [Yes/No]
+    Dependencies:
+
+    Example: private _nodes = [_vehicle] call A3A_fnc_logistics_getVehicleNodes;
+*/
+params ["_vehicle"];
+private _return = -1;
```
Comment (jaj22) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_getVehicleNodes.sqf:20]: Shouldn't this be an empty array to avoid type errors in the calling function?

File: A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf (hunk 8 above)
Comment (HakonRydland) [reply A3-Antistasi/functions/Logistics/fn_logistics_addLoadAction.sqf:29]: think i forgot to test with loading weapons after the change, should definitely not work, fixed

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_getVehicleNodes.sqf (hunk 9 above)
Comment (HakonRydland) [reply A3-Antistasi/functions/Logistics/functions/fn_logistics_getVehicleNodes.sqf:20]: old leftover from JNL, negative int were used a lot to note invalid something, as for the type errors, where its used theres either been confirmed to have valid nodes or theres a type check following it, see logistics_canLoad ln 68.
so it should be fine, unless i missed something

**Comment by HakonRydland** [comment]
> There are two different inputs for A3A_fnc_logistics_addLoadAction
> surrenderAction.sqf
> AIVEHinit.sqf
> createAIAirplane.sqf
> createAIOutpost.sqf
> createAISite.sqf
> use Object call A3A_....
> while
> flagAction.sqf
> AIVEHinit.sqf
> spawnCrate.sqf
> use [object] call A3A_...
> From the looks of it, the array is the right input, but maybe I am missing something

overlooked this comment, sry. you are right it should have an array with a object and optionally a string

**Comment by HakonRydland** [comment]
as a side note, passing a object in place of a array will work fine too as.
[player] params ["_player", "_thisIsNil"];
returns: [player, any]

and
player params ["_player", "_thisIsNil"];
return: [player any]

ie its passed as the first element

**Comment by Bob-Murphy** [comment]
@HakonRydland merge conflicts

**Comment by jaj22** [comment]
Not fully reviewed yet, only did a bit of syntax checking.

**Comment by Bob-Murphy** [comment]
Please create an issue for that PR or make it part of a project.
Thank you.

**Comment by HakonRydland** [comment]
> Please create an issue for that PR or make it part of a project.
> Thank you.

issue created

**Comment by CalebSerafin** [comment]
7 Liner for myself.
* [ ] Indentation & Headers
* [ ] Naming Conventions & Grammar
* [ ] Macros
* [ ] Param & Call
* [ ] Global Variables
* [ ] Local Variables
* [ ] Logic

**Review by CalebSerafin — CHANGES_REQUESTED** [review]
Okay, everything so far looks pretty solid. There are a few comments below that can be changed before I do the logic part of the review. I decided to save the trees, so instead of writing out my comments as a backup, I made an [Excel Spreadsheet](https://1drv.ms/x/s!AhRKsW_EtNcb2lrWnPAuTtV97Tcs?e=15wQ6B).
***
17- I noticed some error codes, could use macros for those. However, this isn't too necessary as we don't have an enumeration system yet.
***
19- Is it possible to prefix logistics_attachmentOffset, logistics_vehicleHardpoints, logistics_weapons, and logistics_coveredVehicles with `A3A_`?
***
The logic review should come in 1-2 days.

File: A3-Antistasi/functions/CREATE/fn_createAISite.sqf [hunk 10]
```diff
@@ -82,7 +82,7 @@ if(_marker in airportsX || {_marker in seaports || {_marker in outposts}}) then
   	_box = CSATAmmoBox createVehicle _markerPos;
```
[the remark this answers is not in the document]
Comment (HakonRydland) [reply A3-Antistasi/functions/CREATE/fn_createAISite.sqf:82]: done, side note: this function is disconnected

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_load.sqf [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    //update list
+    [_vehicle , _node] call _updateList;
+};
+
+//attach data
+private _offsetAndDir = [_cargo] call A3A_fnc_logistics_getCargoOffsetAndDir;
+private _location = _offsetAndDir#0;
+private _location = _location vectorAdd _nodeOffset;
+
+private _bbv = (boundingBoxReal _vehicle select 0 select 1) + ((boundingCenter _vehicle) select 1);
```
[the remark this answers is not in the document]
Comment (HakonRydland) [reply A3-Antistasi/functions/Logistics/functions/fn_logistics_load.sqf:72]: those vowel's are gonna cost you ;)

File: A3-Antistasi/functions/Logistics/fn_logistics_getVehCapacity.sqf [hunk 12]
```diff
@@ -0,0 +1,32 @@
+/*
+    Author: [Håkon]
+    [Description]
+        Returns the vehicles cargo loading capacity
+
+    Arguments:
+    0. <Object> The vehicle you want to know the cargo capacity of
```
[the remark this answers is not in the document]
Comment (HakonRydland) [reply A3-Antistasi/functions/Logistics/fn_logistics_getVehCapacity.sqf:7]: why i like the index of the params in the header, i find its easier to read for potential unexperienced users.

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_canLoad.sqf [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+private _seats = [];
+if ((_node#0) isEqualType []) then {
+    {_seats append (_x#2)} forEach _node;
+} else {
+    _seats append (_node#2);
+};
+private _cargoUnits = [];
+{
+    _x params ["_unit", "_role", "_index", "_turretPath", "_personTurret"];
+    private _isBlocking = call {
```
[the remark this answers is not in the document]
Comment (HakonRydland) [reply A3-Antistasi/functions/Logistics/functions/fn_logistics_canLoad.sqf:97]: none, i probably was exhausted when writing that :)

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_addAction.sqf (hunk 3 above)
Comment (CalebSerafin) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_addAction.sqf:28]: 42- In addition to above
```suggestion
        if (_loadText in _actionNames) exitWith {};
```
This is lightning fast (might be faster than getVariable at 0.8 microseconds). I don't know how SQF implements it in C++.

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_addAction.sqf [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    Environment: Any
+    Public: [No]
+    Dependencies:
+
+    Example: [_object , _action] remoteExec ["A3A_fnc_logistics_addAction", 0, _object];
+*/
+params ["_object", "_action"];
+
+private _actions = [];
+{_actions pushBack (_object actionParams _x)}forEach actionIDs _object;
```
Comment (CalebSerafin) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_addAction.sqf:23]: 41- Could simplfy to the following, which might be slightly faster.
```suggestion
private _actionNames = (actionIDs _object) apply {(_object actionParams _x)#0};
```
In rest of code omit `#0`.

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_canLoad.sqf [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+if ((_node#0) isEqualType []) then {
+    {_seats append (_x#2)} forEach _node;
+} else {
+    _seats append (_node#2);
+};
+private _cargoUnits = [];
+{
+    if ( (_x#2) in _seats ) then {_cargoUnits pushBack _x};
+}forEach _fullCrew;
+if !(_cargoUnits isEqualTo []) exitWith {-9};
```
Comment (CalebSerafin) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_canLoad.sqf:98]: 45- This can be optimised to avoid searching after a sign unit has already being found to be blocking the way.
```suggestion
if !(_fullCrew findIf {_x#2 in _seats} isEqualTo -1) exitWith {-9};
```

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_generateHardPoints.sqf [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+private _radius = (_diameter/2);// get some distance from walls
+private _plane = _planeStart vectorDiff _planeEnd;
+
+private _nodeArray = [];
+private _node = _planeStart vectorAdd [0,_radius,0];//first point
+while {(-_radius*1.5) < (_plane#1)} do {
+    _plane = _plane vectorAdd [0,_diameter,0];
+    _nodeArray pushBack _node;
+    _node = _node vectorAdd [0,_diameter,0];
+};
```
Comment (CalebSerafin) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_generateHardPoints.sqf:52]: 46- Instead of making a negative diametre and radius, we could use vectorDiff.
```suggestion
private _diameter = 0.8;
private _radius = (_diameter/2);// get some distance from walls
private _plane = _planeStart vectorDiff _planeEnd;

private _nodeArray = [];
private _node = _planeStart vectorDiff [0,_radius,0];//first point
while {(_radius*1.5) < (_plane#1)} do {
    _plane = _plane vectorDiff [0,_diameter,0];
    _nodeArray pushBack _node;
    _node = _node vectorDiff [0,_diameter,0];
};
```

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_generateHardPoints.sqf [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+
+//fix nodeArray with rest of information
+private _newArray = [];
+{
+    _newArray pushBack [1, _x, []];
+} forEach _nodeArray;
+A3A_Logistics_nodeArray = _newArray;
```
Comment (CalebSerafin) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_generateHardPoints.sqf:59]: 47- This could be optimised and shortened.
```suggestion
A3A_Logistics_nodeArray = _nodeArray apply {[1, _x, []]};
```

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_initMountedWeapon.sqf [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+    private _vehDir = vectorDir _vehicle;
+    private _weaponDir = _weapon weaponDirection currentWeapon _weapon;
+    private _fireDir = _vehDir vectorAdd (_weaponDir vectorDiff _vehDir);
+
+    private _location = _weapon getVariable ["AttachmentOffset", [0,0,0]];
+
+    private _appliedForce = (_fireDir vectorMultiply -_force);
+    _vehicle addForce [_appliedForce, _location];
+}];
```
Comment (CalebSerafin) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_initMountedWeapon.sqf:42]: 51- It might be better if some values were hard-coded into the delegate so that they do not need to be processed every shot fired.
```suggestion
//credits to audiocustoms on youtube (Cup dev) for the concept and CalebSerafin for optimisation.
private _idRecoil = _weapon addEventHandler ["Fired", compile (' 
    params ["_weapon"];
    private _vehicle = attachedTo _weapon;
    private _weaponDir = _weapon weaponDirection currentWeapon _weapon;
    private _appliedForce = (_weaponDir vectorMultiply -' + (str _fireForce) +');
    _vehicle addForce [_appliedForce, ' + (str (_weapon getVariable ["AttachmentOffset", [0,0,0]])) + '];
')];
```

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_removeWeaponAction.sqf [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+
+//remove Undercover break if last weapon
+private _attachedObjects =  attachedObjects _vehicle;
+private _weaponCount = _attachedObjects findIf {
+    private _model = getText (configFile >> "CfgVehicles" >> typeOf _x >> "model");
+    _weapon = false;
+    {
+        if ((_x#0) isEqualTo _model) exitWith {_weapon = true};
+    } forEach A3A_logistics_weapons;
+    _weapon
```
Comment (CalebSerafin) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_removeWeaponAction.sqf:46]: 52- This could be shortened.
```suggestion
    A3A_logistics_weapons findIf {_x#0 isEqualTo _model} > -1
```

**Review by CalebSerafin — CHANGES_REQUESTED** [review]
Solid code. Comments are mostly optimisation, one array reference possible problem (49).
***
The Excel spreadsheet has been updated to newest comments.

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_getCargoNodeType.sqf [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+*/
+params [["_object", objNull, [objNull, ""]]];
+if (_object isKindOf "CAManBase") exitWith {2};//exception for the medical system
+private _type = if (_object isEqualType objNull) then { typeOf _object } else { _object };
+private _size = -1;
+if (_type isEqualTo "") exitWith {_size};
+
+private _model = getText (configFile >> "CfgVehicles" >> _type >> "model");
+{
+    if ((_x#0) isEqualTo _model) exitWith {_size = +(_x#3)};
```
[the remark this answers is not in the document]
Comment (HakonRydland) [reply A3-Antistasi/functions/Logistics/functions/fn_logistics_getCargoNodeType.sqf:27]: safety, i never use reference to the data structure to ensure its never modified

[the remark this answers is not in the document]
Comment (CalebSerafin) [reply A3-Antistasi/functions/Logistics/functions/fn_logistics_getCargoNodeType.sqf:27]: Is `A3A_logistics_attachmentOffset#any#3` not a scalar for size? All the common basic types (String, Scalar, Boolean,..probably some more but need to test) always copy the value.

File: A3-Antistasi/functions/Logistics/functions/fn_logistics_removeWeaponAction.sqf [hunk 21]
```diff
@@ -39,11 +39,7 @@ _cargo setVariable ["GetOutEH", nil];
 private _attachedObjects =  attachedObjects _vehicle;
 private _weaponCount = _attachedObjects findIf {
     private _model = getText (configFile >> "CfgVehicles" >> typeOf _x >> "model");
-    _weapon = false;
-    {
-        if ((_x#0) isEqualTo _model) exitWith {_weapon = true};
-    } forEach A3A_logistics_weapons;
-    _weapon
+    A3A_logistics_weapons findIf {_x#0 isEqualTo _model} > -1
```
Comment (CalebSerafin) [inline A3-Antistasi/functions/Logistics/functions/fn_logistics_removeWeaponAction.sqf:42]: I think Travis is throwing a hissyfit about this line. It does not need a semi-colon, but maybe add a semi-colon to the end.

Comment (HakonRydland) [reply A3-Antistasi/functions/Logistics/functions/fn_logistics_removeWeaponAction.sqf]: not it apparently

**Review by CalebSerafin — APPROVED** [review]
